sbncode/CAFMaker/CAFMaker_module.cc

```diff
@@ -1010,10 +1013,38 @@ void CAFMaker::produce(art::Event& evt) noexcept {
   GetByLabelStrict(evt, fParams.CRTHitLabel(), crthits_handle);
   // fill into event
   if (crthits_handle.isValid()) {
+
+    //==== gate start time
+    //==== 03/31/22 : 1600000 ns = 1.6 ms is the default T0Offset in MC
+    //==== https://github.com/SBNSoftware/icaruscode/blob/v09_37_02_01/icaruscode/CRT/crtsimmodules_icarus.fcl#L11
+    long long m_gate_start_timestamp = 1600000; // ns
+    if(isRealData){
+
+      art::Handle< std::vector<raw::ExternalTrigger> > externalTrigger_handle;
+      evt.getByLabel( fParams.TriggerLabel(), externalTrigger_handle );
+      const std::vector<raw::ExternalTrigger> &externalTrgs = *externalTrigger_handle;
+
+      art::Handle< std::vector<raw::Trigger> > trigger_handle;
+      evt.getByLabel( fParams.TriggerLabel(), trigger_handle );
+      const std::vector<raw::Trigger> &trgs = *trigger_handle;
+
+      if(externalTrgs.size()==1 && trgs.size()==1){
+        long long TriggerAbsoluteTime = externalTrgs.at(0).GetTrigTime(); // Absolute time of trigger
+        double BeamGateRelativeTime = trgs.at(0).BeamGateTime(); // BeamGate time w.r.t. electronics clock T0 in us
+        double TriggerRelativeTime = trgs.at(0).TriggerTime(); // Trigger time w.r.t. electronics clock T0 in us
+        m_gate_start_timestamp = TriggerAbsoluteTime + (int)(BeamGateRelativeTime*1000-TriggerRelativeTime*1000);
+      }
+      else{
+        std::cerr << "Size of raw::ExternalTrigger = " << externalTrgs.size() << std::endl;
+        std::cerr << "Size of raw::Trigger = " << trgs.size() << std::endl;
+      }
```

Comment:
@jedori0228 has checked the latter suggestion, and for the record, it's wrong. The right one should probably be something like `evt.time().timeHigh() * 1'000'000'000LL + evt.time().timeLow()`, since ICARUS assigns the timestamp not as a single count of nanoseconds from the Epoch (`value()`, 64 bits) but separately as seconds and nanoseconds (`timeHigh()` and `timeLow()` respectively, both 32-bit).

Reply:
As we discusses this in private chat, I found `evt.time().timeHigh() * 1'000'000'000LL + evt.time().timeLow()` is still not identical to `raw::Trigger::TriggerTime()`. We can activate these lines once we understand them better, so I added them but comment out for now.